File: Notes/Computer Science/Algorithms/Meta/Sebastien Rubens Meta Puzzles Solutions/go/l3_slippery_strip/l3_slippery_strip.go
// meta_puzzles by Sebastien Rubens
//
// Please go to https://github.com/seb-pg/meta_puzzles/README.md
// for more information
//
// To the extent possible under law, the person who associated CC0 with
// meta_puzzles has waived all copyright and related or neighboring rights
// to meta_puzzles.
//
// You should have received a copy of the CC0 legalcode along with this
// work.  If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.

package l3_slippery_strip

import (
	"meta_puzzles/std"
	"meta_puzzles/test"
)

type row_t = []byte
type char_counter_t = [256]int

type Counts struct {
	star  int
	right int
	down  int
}

type predicate_t[T any] func(x T) bool

func Count[T std.Comparable](elements []T, value T) int {
	// similar to C++ std::count
	nb := 0
	for _, elt := range elements {
		if elt == value {
			nb += 1
		}
	}
	return nb
}

func _find_element[T any](elements []T, pred predicate_t[T]) int {
	for pos, elt := range elements {
		if pred(elt) {
			return pos
		}
	}
	return len(elements)
}

func Find[T std.Comparable](elements []T, value T) int {
	// similar to C++ std::find
	fn := func(x T) bool { return x == value }
	return _find_element(elements, fn)
}

func get_counts(row row_t, counts *char_counter_t) Counts {
	counts['*'] = 0
	counts['>'] = 0
	counts['v'] = 0
	for _, c := range row {
		d := int(c)
		if d >= len(counts) {
			d = '.' // empty value (we do not use it)
		}
		counts[d] += 1
	}
	return Counts{counts['*'], counts['>'], counts['v']}
}

func get_nb_coins_right_then_down3(row row_t, count_down int, count_right int) int {
	// complexity : O(C)
	if count_down == 0 {
		return 0
	}
	k := Find(row, 'v')
	j := 0
	if k != len(row) {
		j = k + 1
	}
	new_row := make(row_t, 0, len(row))
	new_row = append(new_row, row[j:]...)
	new_row = append(new_row, row[:j]...)
	nb_coins_right_then_down := 0
	last := 0
	for count_right*count_down != 0 {
		first := last + Find(new_row[last:], '>')
		last := first + Find(new_row[first:], 'v')
		nb_coins_right_then_down = std.Max(nb_coins_right_then_down, Count(new_row[first:last], '*'))
		count_down -= 1
		count_right -= Count(new_row[first:last], '>')
	}
	return nb_coins_right_then_down
}

func getMaxCollectableCoins(R int32, C int32, G [][]byte) int32 {
	counts := char_counter_t{}
	res := 0
	for i, _ := range G {
		row := G[len(G)-i-1]
		count := get_counts(row, &counts)
		nb_coins_immediately_down := std.Min(count.star, 1)
		if count.right == int(C) {
			res = 0
		} else if count.right == 0 {
			res += nb_coins_immediately_down
		} else {
			nb_coins_right_then_down := get_nb_coins_right_then_down3(row, count.down, count.right)
			nb_coins_right_forever := 0
			if count.down == 0 {
				nb_coins_right_forever = count.star
			}
			res = std.Max(nb_coins_immediately_down+res, std.Max(nb_coins_right_then_down+res, nb_coins_right_forever))
		}
	}
	return int32(res) // result should always be positive
}

type Args struct {
	G   []string
	res int32
}

func (self Args) GetResult() any {
	return self.res
}

func Tests() uint {
	wrapper := func(p Args) int32 {
		H := make([][]byte, len(p.G))
		for idx, row := range p.G {
			H[idx] = []byte(string(row))
		}
		C := 0
		if len(H) > 0 {
			C = len(H[0])
		}
		return getMaxCollectableCoins(int32(len(H)), int32(C), H)
	}

	args_lists := []Args{
		{[]string{".***", "**v>", ".*.."}, 4},
		{[]string{">**", "*>*", "**>"}, 4},
		{[]string{">>", "**"}, 0},
		{[]string{">*v*>*", "*v*v>*", ".*>..*", ".*..*v"}, 6},
		// extra1
		{[]string{}, 0},
		{[]string{"."}, 0},
		{[]string{"v"}, 0},
		{[]string{">"}, 0},
		{[]string{"*"}, 1},
		// extra2
		{[]string{".", ".", ">", "*"}, 0},
		{[]string{".", "*", ">", "*"}, 1},
		{[]string{".", "*", ">", "."}, 1},
		{[]string{"*", ".", ">", "."}, 1},
		{[]string{"***", "...", ">vv", "..."}, 1},
		// extra3
		{[]string{"*....", ".*...", "..*..", "...*."}, 4},
		{[]string{"....", "....", "....", "...."}, 0},
		{[]string{"***>", "...."}, 3},
		{[]string{"...."}, 0},
		{[]string{"vvvv"}, 0},
		{[]string{"vvvv", "....", ">>>>"}, 0},
		// extra4
		{[]string{"******", "......", ">*>vvv", "......"}, 2},
		{[]string{"*****", ".....", ">>vvv", "....."}, 1},
	}

	return test.RunAllTests("l3_slippery_strip", args_lists, wrapper)
}

File: Notes/Computer Science/Algorithms/Meta/Sebastien Rubens Meta Puzzles Solutions/cpp17/l3_slippery_strip.cpp
// meta_puzzles by Sebastien Rubens
//
// Please go to https://github.com/seb-pg/meta_puzzles/README.md
// for more information
//
// To the extent possible under law, the person who associated CC0 with
// meta_puzzles has waived all copyright and related or neighboring rights
// to meta_puzzles.
//
// You should have received a copy of the CC0 legalcode along with this
// work.  If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.

#include "test_all.h"

#include <algorithm>
#include <array>
#include <vector>

namespace l3_slippery_strip {

struct Counts
{
    uint32_t star;
    uint32_t right;
    uint32_t down;
};

using row_t = std::vector<char>;
using char_counter_t = std::array<uint32_t, 256>;

static Counts get_counts(const row_t& row, char_counter_t& counts)
{
    counts['*'] = 0;
    counts['>'] = 0;
    counts['v'] = 0;
    for (auto c : row)
    {
        auto d = static_cast<size_t>(c);
        if (d >= std::size(counts))
            d = '.';  // empty value (we do not use it)
        ++counts[d];
    }
    return { counts['*'], counts['>'], counts['v'] };
}

static uint32_t get_nb_coins_right_then_down3(const row_t& row, uint32_t count_down, uint32_t count_right)
{
    // complexity : O(C)
    if (count_down == 0)
        return 0;
    const auto k = std::find(std::cbegin(row), std::cend(row), 'v');
    const auto j = k != std::cend(row) ? k + 1 : std::cbegin(row);
    row_t new_row;
    new_row.insert(std::end(new_row), j, std::cend(row));
    new_row.insert(std::end(new_row), std::cbegin(row), j);
    uint32_t nb_coins_right_then_down = 0;
    auto last = std::cbegin(new_row);
    while (count_right * count_down != 0)
    {
        const auto first = std::find(last, std::cend(new_row), '>');
        last = std::find(first, std::cend(new_row), 'v');
        nb_coins_right_then_down = std::max(nb_coins_right_then_down, static_cast<uint32_t>(std::count(first, last, '*')));
        --count_down;
        count_right -= static_cast<int32_t>(std::count(first, last, '>'));
    }
    return nb_coins_right_then_down;
}

int32_t getMaxCollectableCoinsCpp17(uint32_t R, uint32_t C, const std::vector<std::vector<char>>& G)
{
    // https://www.metacareers.com/profile/coding_puzzles/?puzzle=2881982598796847
    // Constraints
    //      2 ≤ R, C ≤ 400, 000
    //      R∗C ≤ 800, 000
    //      Gi, j ∈{ ".", "*", ">", "v" }
    //      Complexity: O(N), where N = R * C

    (void)R;

    char_counter_t counts;
    std::fill(std::begin(counts), std::end(counts), 0);
    uint32_t res = 0;
    for (size_t i = 0, e = G.size(); i < e; ++i)
    {
        const auto& row = G[G.size() - i - 1];
        const auto count = get_counts(row, counts);
        const auto nb_coins_immediately_down = std::min(count.star, 1u);
        if (count.right == C)
            res = 0;
        else if (count.right == 0)
            res += nb_coins_immediately_down;
        else
        {
            const auto nb_coins_right_then_down = get_nb_coins_right_then_down3(row, count.down, count.right);
            const auto nb_coins_right_forever = count.down == 0 ? count.star : 0;
            res = std::max(nb_coins_immediately_down + res, std::max(nb_coins_right_then_down + res, nb_coins_right_forever));
        }
    }
    return static_cast<int32_t>(res);  // result should always be positive
}

using namespace std;

int getMaxCollectableCoins(int R, int C, vector<vector<char>> G) {
    return getMaxCollectableCoinsCpp17(static_cast<uint32_t>(R), static_cast<uint32_t>(C), G);
}


struct Args
{
    std::vector<std::string> G;
};

auto tests()
{
    const auto wrapper = [](Args& p)
    {
        std::vector<std::vector<char>> H;
        H.reserve(p.G.size());
        for (const auto& row : p.G)
        {
            auto& last = H.emplace_back();
            last.insert(last.end(), std::cbegin(row), std::cend(row));
        }
        return getMaxCollectableCoins(static_cast<int>(H.size()), H.empty() ? 0 : static_cast<int>(H[0].size()), H);
    };

    std::vector<NamedTests<Args, int>> tests = {
        { "Meta", {
                { { { ".***", "**v>", ".*.." } }, 4 },
                { { { ">**", "*>*", "**>" } }, 4 },
                { { { ">>", "**" } }, 0 },
                { { { ">*v*>*", "*v*v>*", ".*>..*", ".*..*v" } }, 6 },
            }
        },
        {"extra1", {
                { { { } }, 0 },
                { { { "." } }, 0 },
                { { { "v" } }, 0 },
                { { { ">" } }, 0 },
                { { { "*" } }, 1 },
            }
        },
        { "extra2", {
                { { { ".", ".", ">", "*" } }, 0 },
                { { { ".", "*", ">", "*" } }, 1 },
                { { { ".", "*", ">", "." } }, 1 },
                { { { "*", ".", ">", "." } }, 1 },
                { { { "***", "...", ">vv", "..." } }, 1 },
            }
        },
        { "extra3", {
                { { { "*....", ".*...", "..*..", "...*." } }, 4 },
                { { { "....", "....", "....", "...." } }, 0 },
                { { { "***>", "...." } }, 3 },
                { { { "...." } }, 0 },
                { { { "vvvv" } }, 0 },
                { { { "vvvv", "....", ">>>>" } }, 0 },
            }
        },
        { "extra4", {
                { { { "******", "......", ">*>vvv", "......" } }, 2 },
                { { { "*****", ".....", ">>vvv", "....." } }, 1 },
            }
        },
    };

    return run_all_tests("l3_slippery_strip", tests, wrapper);
}

}  // namespace l3_slippery_strip
